File: 0_5_DFS.java
import java.util.*;
public class DFS {
    // attributes for this class
    // vertices
    private int v;
    // adjacency list
    private List<Integer> adjList[];
    // visited  to check a vertex is visited or not in Explore method
    private boolean visited[];
    // for the preOrder traversal
    private int pre[];
    // counter for the preorder traversal
    private int counterPre = 0;

    // for assigning all the pre defined data structure
    DFS(int ver){
        // assign the nu,mber of vertex
        v = ver;
        // assign the number of boolean at it's vertex
        visited = new boolean[ver];
        // assisgn the array with the vertex length
        pre = new int[ver];
        // asssigning the adjList length
        adjList = new ArrayList[ver];

        // now going throgh the loop and adding the newList to adjList each index , in pre making all -1, and false in all visited
        for (int x=0; x<ver; x+=1){
            adjList[x] = new ArrayList();

            // for the visited boolean value false
            visited[x] = false;

            // pre 
            pre[x]=-1;
        }
    }

    // method to add the edge
    // as it is undirected we have to add n both direction
    void addEdge(int vertex1, int vertex2){
        adjList[vertex1].add(vertex2);
        adjList[vertex2].add(vertex1);
    }
    
    // for exploring the vertex
    void Explore(int vertex){
        // first mark it as true as it has been visited
        visited[vertex]=true;

        //now make the pre at the counterPre index this vertex
        pre[counterPre] = vertex;
        // also increase the counterPre for next iteration
        counterPre+=1;

        // do exploration
        // using Iterator to iterate the adjList
        Iterator<Integer> i = adjList[vertex].listIterator();
        while(i.hasNext()){
            int ver = i.next();

            // check is it visited or not if not then again explore that ver
            if (visited[ver]==false){
                Explore(ver);
            }
        }
    }

    // dfs depth forst search
    void dfs(){
        // start from the index 0
        for (int vertex=0; vertex<v; vertex++){
            // check visited or not
            if (visited[vertex]==false){
                Explore(vertex);
            }
        }
    }
    public static void main(String[] args){
        //  for scanning
        Scanner scanner = new Scanner(System.in);

        int ver = scanner.nextInt();// having the vertex number
        int edge = scanner.nextInt();// having the edge number

        DFS graph = new DFS(ver);// defing the graph 

        // now make the edges as adjacency List
        for (int x=0; x< edge; x++){
            int vertex1= scanner.nextInt();
            int vertex2= scanner.nextInt();

            // add the edge
            graph.addEdge(vertex1, vertex2);
        }
        scanner.close();
        
        // now do depth first search
        graph.dfs();

        for(int x=0; x<ver;x++) System.out.print(graph.pre[x]+"\t");
    }
}

File: 0_5_DFS.py
import threading
class Graph:
    def __init__(self, ver):
        # number of vertex
        self.vertex = ver
        # adjacency list
        self.adjList = [[] for x in range(ver)]
        # for marking visited or not during exploring vertices
        self.visited = [False for x in range(ver)]
        # add the element as visited next
        self.pre = [-1 for x in range(ver)]
        #  keep the counter
        self.countpre = 0

    # explore that particular vertex ver
    def Explore(self, ver):
        # mark it as the visited
        self.visited[ver]=True

        # for exploring in sequence  --> preorder
        self.pre[self.countpre] = ver
        self.countpre+=1# increase fro next iteration


        # explore if not visited inside that adjacency list of vertex ver
        for vertex in self.adjList[ver]:
            if self.visited[vertex]==False:
                self.Explore(vertex)

    # for the depth first search
    def DFS(self):
        # traverse through all the vertices
        for ver in range(self.vertex):
            # if not visited then explore it
            if self.visited[ver]==False:
                self.Explore(ver)


def main():
    vertex, edge = [int(x) for x in input('vertex and edges --> \t').split()]

    graph = Graph(vertex)

    for x in range(edge):
        ver1, ver2 = [int(x) for x in input().split()]

        # as it is undirected so we have to add it as the both direction 
        graph.adjList[ver1].append(ver2)
        graph.adjList[ver2].append(ver1)

    # now have the dfs
    graph.DFS()

    # print the explore vertices in te sequence
    print(*graph.pre)

threading.Thread(target=main).start()

'''
input
4 6
0 1
0 2
1 2
2 0
2 3
3 3

output
0 1 2 3



input
6 6
2 3
3 2
2 1
1 2
0 4
5 2
output
0 4 1 2 3 5


input
9 8
0 1
0 2
0 4
2 3
5 0
6 7
8 7
8 6

output
0 1 2 3 4 5 6 7 8
'''
